### ISSUE TITLE
Compiler bug when generating tests with macros.

### ISSUE THREAD

**Issue by lobachevzky** [reem/stainless#76]
I found a compiler bug using stainless. I've also posted this on rust repo.

The simplest version of the code that produces the bug is this:
```Rust
describe! tests {
    before_each {
        macro_rules! make_test {
        }
    }

    make_test!();

    after_each {
    }
}
```

The bug that it throws is:
```
   Fresh stainless v0.1.10
   Compiling pipes v0.1.0 (file:///home/ethanbro/lstm)
   Running `rustc main.rs --crate-name pipes -g --test -C metadata=b8dfe5ce564a1ec9 -C extra-filename=-b8dfe5ce564a1ec9 --out-dir /home/ethanbro/lstm/target/debug/deps --emit=dep-info,link -L dependency=/home/ethanbro/lstm/target/debug/deps --extern stainless=/home/ethanbro/lstm/target/debug/deps/libstainless-195115ceaf447e08.so`
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Diagnostic { level: Fatal, message: "Expected one of: `before_each, after_each, before, after, it, bench, failing, describe, ignore`, but found: `make_test`", code: None, span: MultiSpan { primary_spans: [Span { lo: BytePos(3788), hi: BytePos(3789), expn_id: ExpnId(4294967295) }], span_labels: [] }, children: [] }', /home/ethanbro/.cargo/registry/src/github.com-1ecc6299db9ec823/stainless-0.1.10/src/parse.rs:192
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `pipes`.

Caused by:
  process didn't exit successfully: `rustc main.rs --crate-name pipes -g --test -C metadata=b8dfe5ce564a1ec9 -C extra-filename=-b8dfe5ce564a1ec9 --out-dir /home/ethanbro/lstm/target/debug/deps --emit=dep-info,link -L dependency=/home/ethanbro/lstm/target/debug/deps --extern stainless=/home/ethanbro/lstm/target/debug/deps/libstainless-195115ceaf447e08.so` (exit code: 101)

shell returned 101
```

Let me know if I can provide any additional info.

**Comment by ujh** [comment]
Unfortunately I think I'll have to close this. Stainless is a giant hack as it just transforms the code down to the builtin test support that comes with Rust. That has a lot of weird side effects (like before_each not working in benchmarks, etc.). IMO the way forward is to built a proper XUnit testing framework complete with a custom test runner. I [started exploring this](https://github.com/ujh/yatf) but I haven't gotten far yet. Or rather I haven't put in the time to get this to work.

**Comment by utkarshkukreti** [comment]
> That has a lot of weird side effects (like before_each not working in benchmarks, etc.)

I'd just like to point out that `before_each` can be made to work with `#[bench]`. I've done it in my test framework that was heavily inspired by stainless: https://github.com/utkarshkukreti/speculate.rs/blob/master/benches/basic.rs I created a new crate because I wanted to include some breaking changes like string descriptions for describe blocks. It also supports nested before/after blocks and I've been using it in several projects for over 2 years now (I used stainless before that).
